packages/netsuite-adapter/src/mapped_lists/utils.ts

```diff
@@ -0,0 +1,228 @@
+
+const { awu } = collections.asynciterable
+const { makeArray } = collections.array
+
+const log = logger(module)
+
+export type MappedList = {
+  path: ElemID
+  value: Values
+}
+
+export const validateTypesFieldMapping = (elements: ObjectType[]): void => {
+  const typesMap = _.keyBy(elements, element => element.elemID.name)
+  Object.entries(listMappedByFieldMapping).forEach(([typeName, fieldNameList]) => {
+    const type = typesMap[typeName]
+    if (type === undefined) {
+      log.error(`type '${typeName}' should have mapping field but '${typeName}' is not in elements`)
+      throw new Error('missing some types with field mapping')
+    }
+    makeArray(fieldNameList).forEach(fieldName => {
+      if (type.fields[fieldName] === undefined) {
+        log.error(`type '${typeName}' should have mapping field '${fieldName}' but '${typeName}' has no '${fieldName}' field`)
+        throw new Error('missing some types with field mapping')
+      }
+    })
+  })
+}
+
+export const convertFieldsTypesFromListToMap = async (
+  element: ObjectType,
+): Promise<void> => {
+  await awu(Object.entries(element.fields)).forEach(async ([fieldName, field]) => {
+    const fieldType = await field.getType()
+    if (!isListType(fieldType)) {
+      return
+    }
+
+    const innerType = await fieldType.getInnerType()
+    if (!isObjectType(innerType)) {
+      return
+    }
+
+    const listMappedByField = isField(innerType.fields.scriptid)
+      ? SCRIPT_ID
+      : listMappedByFieldMapping[innerType.elemID.name]
+    if (listMappedByField === undefined) {
+      log.warn(`there's no field to map by in ${innerType.elemID.getFullName()}`)
+      return
+    }
+
+    const newField = new Field(
+      field.parent,
+      field.name,
+      createRefToElmWithValue(new MapType(innerType)),
+      { ...field.annotations, [LIST_MAPPED_BY_FIELD]: listMappedByField }
+    )
+    element.fields[fieldName] = newField
+    innerType.fields[INDEX] = new Field(
+      innerType,
+      INDEX,
+      createRefToElmWithValue(BuiltinTypes.NUMBER)
+    )
+  })
+}
+
+const getItemKey = (mapFieldValue: string, path?: ElemID): string => {
+  const naclCaseKey = naclCase(mapFieldValue)
+  if (mapFieldValue === naclCaseKey) {
+    return mapFieldValue
+  }
+
+  const serviceIdInfoList = captureServiceIdInfo(mapFieldValue)
+  if (serviceIdInfoList.length === 0) {
+    log.debug(`fixing mapFieldValue to use it as a key: '${mapFieldValue}' -> '${naclCaseKey}' (${path?.getFullName()})`)
+    return naclCaseKey
+  }
+
+  const serviceIdKey = naclCase(
+    serviceIdInfoList.map(serviceIdInfo => serviceIdInfo[CAPTURED_SERVICE_ID]).join('_')
+  )
+  log.debug(`extracting ${SCRIPT_ID} to use it as a key: '${mapFieldValue}' -> '${serviceIdKey}' (${path?.getFullName()})`)
+  return serviceIdKey
+}
+
+const addSuffixUntilUnique = (
+  keyName: string,
+  keySet: Set<string>,
+  path?: ElemID,
+  suffixIndex = 0,
```

Comment:
IMO it will be more readable if we would start from 2
```suggestion
  suffixIndex = 2,
```